### PR TITLE
Issue 22882 get save contentlet version info using variant

### LINKED ISSUE #22882: Get/Save ContentletVersionInfo using Variant
Get Using variant:

- [x]  Change to look for variant, we need to overwrite this method one with two parameters and look for Default variant and another one to receive 3 parameters and look for any variant

https://github.com/dotCMS/core/blob/6b23b2df78e7856c8c960d3e63863bbf8fa5c11b/dotCMS/src/main/java/com/dotmarketing/business/VersionableFactoryImpl.java#L352

- [x] Change cache key to include variant

https://github.com/dotCMS/core/blob/cfc0fe8893ddf426ec0eafc034e525782d300124/dotCMS/src/main/java/com/dotmarketing/business/IdentifierCacheImpl.java#L240

**Update VersionableFactoryImpl**

- [x]  Update these...

https://github.com/dotCMS/core/blob/6b23b2df78e7856c8c960d3e63863bbf8fa5c11b/dotCMS/src/main/java/com/dotmarketing/business/VersionableFactoryImpl.java#L445

https://github.com/dotCMS/core/blob/6b23b2df78e7856c8c960d3e63863bbf8fa5c11b/dotCMS/src/main/java/com/dotmarketing/business/VersionableFactoryImpl.java#L419

...to include variant

- [x] Also we need to update the contentlet editor to send the variant that we are working on every time a contentlet is saved, for now we are not going to work with a current variant in session because we don't need it to Experiment so we are going the current variant from the request as query params and if this query params is not sent we are going to use the default variant.

- [x] Create a Task Upgrade to add the variant_id field into the primary key in contentlet_version_info

https://github.com/dotCMS/core/blob/5fa78680958fe4a80d4a7a9a6feadef6bfc3ac65/dotCMS/src/main/resources/postgres.sql#L564

- [x] Make sure the nothing is broken with these changes, important feature to check are: Push Publish (don't forget Static Push), Time Machine, Site Search, CRUD in contenlet portlet, page render.

### PR DESCRIPTION
**Pull request by freddyucv**
Change to allow save and get ContentletVersionInfo by variant, how is describe here:

https://github.com/dotCMS/core/issues/22882

Also changes to save the variant into the contentlet like:

- https://github.com/dotCMS/core/pull/22933/files#diff-4c3a157c1ca3688be7680e3b7eefa73208cc44d04de7d280d335427c6f300c39R164
- https://github.com/dotCMS/core/pull/22933/files#diff-8ff85036abfa20136fa1fb6ef727e57ed9d05a10338225f75ea92513a5fefa67R57

and to look for variant to know whether we need to create a new inode or not

- https://github.com/dotCMS/core/pull/22933/files#diff-6bf601075adc9b86a64f57ca55728b7bae35c56dcd5896c7f5c05b3cb46fcc8aR365
- [getContentletVersionInfo](https://github.com/dotCMS/core/pull/22933/files#diff-b5d82de048877f797337b47640a36eadbde435004223134e7301ee64ebe9ededR1045)
- https://github.com/dotCMS/core/pull/22933/files#diff-fa1ceaa19618a6b2bbc30e24c6f930b4971f417db50babb748c2e2837ba9eb82R8472

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java [hunk 1]
```diff
@@ -95,6 +96,7 @@ public class Contentlet implements Serializable, Permissionable, Categorizable,
   public static final String HAS_TITLE_IMAGE_KEY = "hasTitleImage";
   public static final String INODE_KEY = "inode";
   public static final String LANGUAGEID_KEY = "languageId";
+  public static final String VARIANT_ID = "VariantId";
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java:99]: Change value to "variantId"

File: dotCMS/src/main/java/com/dotmarketing/business/ContentletVersionInfoTransformer.java [hunk 2]
```diff
@@ -43,6 +46,12 @@ private static ContentletVersionInfo transform(Map<String, Object> map)  {
         versionInfo.setLockedBy((String) map.get("locked_by"));
         versionInfo.setLockedOn((Date) map.get("locked_on"));
         versionInfo.setVersionTs((Date) map.get("version_ts"));
+
+        final String variantId = UtilMethods.isSet(map.get(Contentlet.VARIANT_ID)) ?
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/business/ContentletVersionInfoTransformer.java:50]: or do
map.getOrDefault(Contentlet.VARIANT_ID, VariantAPI.DEFAULT_VARIANT.identifier())

File: dotCMS/src/main/java/com/dotmarketing/business/VersionableFactoryImpl.java [hunk 3]
```diff
@@ -322,43 +323,59 @@ protected void saveVersionInfo(VersionInfo info, boolean updateVersionTS) throws
...
+    protected Optional<ContentletVersionInfo> getContentletVersionInfo(final String identifier,
+			final long lang) throws DotDataException, DotStateException {
+        return getContentletVersionInfo(identifier, lang, VariantAPI.DEFAULT_VARIANT.identifier());
+    }
+
+	@Override
+	public Optional<ContentletVersionInfo> getContentletVersionInfo(
+			final String identifier, final long lang, final String variantId) throws DotDataException, DotStateException{
+
+		if (DbConnectionFactory.inTransaction()) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/VersionableFactoryImpl.java:335]: this is very good pattern

File: dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java [hunk 4]
```diff
@@ -1618,8 +1620,15 @@ public ManifestInfo getManifestInfo() {
 
 	}
 
+    public void setVariantId(final String variantId) {
+		map.put(VARIANT_ID, variantId);
+    }
+	public String getVariantId() {
+		return UtilMethods.isSet(map.get(VARIANT_ID)) ? map.get(VARIANT_ID).toString() :
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java:1627]: map.getOrDefault(VARIANT_ID, VariantAPI.DEFAULT_VARIANT.identifier());

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotmarketing/business/VersionableFactoryImplTest.java [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+        final Language language = new LanguageDataGen().nextPersisted();
+        final ContentType contentType = new ContentTypeDataGen().nextPersisted();
+        final Contentlet contentlet = new ContentletDataGen(contentType).nextPersisted();
+
+        final Identifier identifier = APILocator.getIdentifierAPI()
+                .find(contentlet.getIdentifier());
+
+        FactoryLocator.getVersionableFactory()
+                .createContentletVersionInfo(identifier, language.getId(), contentlet.getInode());
```
[the remark this answers is not in the document]
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotmarketing/business/VersionableFactoryImplTest.java:56]: done https://github.com/dotCMS/core/pull/22933/commits/96edb299e4be6a1b653d61e3b97f90c443ce9061#diff-375a80e44b2e95ca01d97331463781560625a991f95daaaa2bf6afc014dda559R58

File: dotCMS/src/main/java/com/dotmarketing/business/ContentletVersionInfoTransformer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotmarketing/business/ContentletVersionInfoTransformer.java]: done https://github.com/dotCMS/core/pull/22933/commits/96edb299e4be6a1b653d61e3b97f90c443ce9061#diff-406d3ab5ccbc978dfca4801238a1bf140aa778586be51c16cd0d3489e92607b8R49

File: dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java (hunk 1 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java]: done https://github.com/dotCMS/core/pull/22933/commits/96edb299e4be6a1b653d61e3b97f90c443ce9061#diff-1367236d40a612b923ececfa521053143f026f60f035c9f9963aae4efdfaf034R99

File: dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java (hunk 4 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java]: done https://github.com/dotCMS/core/pull/22933/commits/225490b2fcd2f80ffc3fc7bc47a3429e1db6789f#diff-1367236d40a612b923ececfa521053143f026f60f035c9f9963aae4efdfaf034R1627

**Review by dsilvam — APPROVED** [review]

**Pull request merged**